platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/orphan/OrphanBuffer.java

```diff
@@ -0,0 +1,254 @@
+        missingParentMap.shiftWindow(
+                minimumGenerationNonAncient,
+                (parent, orphans) -> ancientParents.add(new ParentAndOrphans(parent, orphans)));
+
+        ancientParents.forEach(this::missingParentBecameAncient);
+    }
+
+    /**
+     * Called when a parent becomes ancient.
+     * <p>
+     * Accounts for events potentially becoming un-orphaned as a result of the parent becoming ancient.
+     *
+     * @param parentAndOrphans the parent that became ancient, along with its orphans
+     */
+    private void missingParentBecameAncient(@NonNull final ParentAndOrphans parentAndOrphans) {
+        final EventDescriptor parentDescriptor = parentAndOrphans.parent();
+
+        for (final OrphanedEvent orphan : parentAndOrphans.orphans()) {
+            orphan.missingParents().remove(parentDescriptor);
+
+            if (orphan.missingParents().isEmpty()) {
+                eventIsNotAnOrphan(orphan.orphan());
+            }
+        }
+    }
+
+    /**
+     * Get the parents of an event that are currently missing.
+     *
+     * @param event the event whose missing parents to find
+     * @return the list of missing parents, empty if no parents are missing
+     */
+    @NonNull
+    private List<EventDescriptor> getMissingParents(@NonNull final GossipEvent event) {
+        final List<EventDescriptor> missingParents = new ArrayList<>();
+
+        final Iterator<EventDescriptor> parentIterator = new ParentIterator(event);
+        while (parentIterator.hasNext()) {
+            final EventDescriptor parent = parentIterator.next();
+            if (!eventsWithParents.contains(parent) && parent.getGeneration() >= minimumGenerationNonAncient) {
+                missingParents.add(parent);
+            }
+        }
+
+        return missingParents;
+    }
+
+    /**
+     * Signal that an event is not an orphan.
+     * <p>
+     * Accounts for events potentially becoming un-orphaned as a result of this event not being an orphan.
+     *
+     * @param event the event that is not an orphan
+     */
+    private void eventIsNotAnOrphan(@NonNull final GossipEvent event) {
+        final Deque<GossipEvent> nonOrphanStack = new LinkedList<>();
+        nonOrphanStack.push(event);
+
+        while (!nonOrphanStack.isEmpty()) {
+            final GossipEvent nonOrphan = nonOrphanStack.pop();
+            final EventDescriptor nonOrphanDescriptor = nonOrphan.getDescriptor();
+
+            if (nonOrphan.getGeneration() < minimumGenerationNonAncient) {
+                // Although it doesn't cause harm to pass along ancient events, it is unnecessary to do so.
+                intakeEventCounter.eventExitedIntakePipeline(event.getSenderId());
+                currentOrphanCount--;
+                continue;
+            }
+
+            eventConsumer.accept(nonOrphan);
+            eventsWithParents.add(nonOrphanDescriptor);
+            currentOrphanCount--;
```

Comment:
If you move `currentOrphanCount--` to the top of the while loop and remove it from the current locations, the code is logically equivalent I think.